README
ALPHA - WORK IN PROGRESS
    *alpha* not recommended for anything but hacking on some simple CLucene
    examples

NAME
    SHEEP - simple CLucene wrapper that spawns threads for every shard when
    search()ing

SYNOPSIS
      use SHEEP;
      my $x = new SHEEP('/path/to/index',5) # create index with 5 shards

      my @documents = ();
      push @documents, { "name" => "jack doe", age => "200" }
          for(1 .. 1_000);

      push @documents, { "name" => "john doe", age => "300" }
          for(1 .. 1_000);
      $x->index(\@documents);

      # find top 10 documents containing 'doe' in the name field
      my $r = $x->search({ term => { name => 'doe' } }, 10); 

      # find top 10 documents containing 'doe' in the name,
      # AND 200 in the 'age' field
      my $r = $x->search({ term => { name => 'doe', age => '200' } }, 10); 

      # some bool query examples:
      $x->search({ 
                    bool => {
                        must => [
                           { term => { "name" => 'jack' } }
                        ],
                        must_not => [
                           { term => { "name" => 'john' } }
                        ]
                    }
                 }, 10);

      $x->search({ 
                    bool => {
                        must => [
                           { term => { "name" => 'jack' } }
                        ],
                        must_not => [
                           { term => { "name" => 'jack' } }
                        ]
                    }
                 }, 10);

      $x->search({ 
                    bool => {
                        must => [
                           { term => { "name" => 'jack' } }
                        ],
                        should => [
                           { term => { "name" => 'jim' } }
                        ],
                        must_not => [
                           { term => { "name" => 'john' } }
                        ]
                    }
                 }, 10);

      $x->search({ 
                    bool => {
                        must => [
                           { 
                               bool => {  
                                   must => [ { term => { name => 'jack' } } ]
                               }
                           }
                        ],
                        should => [
                           { term => { name => 'jim' } }
                        ],
                        must_not => [
                           { term => { name => 'john' } }
                        ]
                    }
                 }, 10);

DESCRIPTION
    SHEEP is a simplistic non-flexible CLucene wrapper

    It uses WhiteSpaceAnalyzer, and can work only with BooleanQueries and
    Term queries. It also uses boost::thread to spawn new thread for every
    shard when searching.

SEE ALSO
    <http://clucene.sourceforge.net/>

    <http://clucene.sourceforge.net/doc/html/namespaces.html>

    <https://github.com/jackdoe/SHEEP>

AUTHOR
    Borislav Nikolov <jack@sofialondonmoskva.com>

COPYRIGHT AND LICENSE
    Copyright (C) 2013 by Borislav Nikolov

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.18.1 or, at
    your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension SHEEP.

0.01  Sat Dec 28 10:08:28 2013
	- original version; created by h2xs 1.23 with options
		-A -n SHEEP



SHEEP.c
/*
 * This file was generated automatically by ExtUtils::ParseXS version 3.18 from the
 * contents of SHEEP.xs. Do not edit this file, edit SHEEP.xs instead.
 *
 *    ANY CHANGES MADE HERE WILL BE LOST!
 *
 */

#line 1 "SHEEP.xs"
#ifdef __cplusplus
extern "C" {
#endif

#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"
#include "ppport.h"
#undef do_open
#undef do_close

#ifdef __cplusplus
}
#endif

#include "cpp/Sheep.hpp"

#line 28 "SHEEP.c"
#ifndef PERL_UNUSED_VAR
#  define PERL_UNUSED_VAR(var) if (0) var = var
#endif

#ifndef dVAR
#  define dVAR		dNOOP
#endif


/* This stuff is not part of the API! You have been warned. */
#ifndef PERL_VERSION_DECIMAL
#  define PERL_VERSION_DECIMAL(r,v,s) (r*1000000 + v*1000 + s)
#endif
#ifndef PERL_DECIMAL_VERSION
#  define PERL_DECIMAL_VERSION \
	  PERL_VERSION_DECIMAL(PERL_REVISION,PERL_VERSION,PERL_SUBVERSION)
#endif
#ifndef PERL_VERSION_GE
#  define PERL_VERSION_GE(r,v,s) \
	  (PERL_DECIMAL_VERSION >= PERL_VERSION_DECIMAL(r,v,s))
#endif
#ifndef PERL_VERSION_LE
#  define PERL_VERSION_LE(r,v,s) \
	  (PERL_DECIMAL_VERSION <= PERL_VERSION_DECIMAL(r,v,s))
#endif

/* XS_INTERNAL is the explicit static-linkage variant of the default
 * XS macro.
 *
 * XS_EXTERNAL is the same as XS_INTERNAL except it does not include
 * "STATIC", ie. it exports XSUB symbols. You probably don't want that
 * for anything but the BOOT XSUB.
 *
 * See XSUB.h in core!
 */


/* TODO: This might be compatible further back than 5.10.0. */
#if PERL_VERSION_GE(5, 10, 0) && PERL_VERSION_LE(5, 15, 1)
#  undef XS_EXTERNAL
#  undef XS_INTERNAL
#  if defined(__CYGWIN__) && defined(USE_DYNAMIC_LOADING)
#    define XS_EXTERNAL(name) __declspec(dllexport) XSPROTO(name)
#    define XS_INTERNAL(name) STATIC XSPROTO(name)
#  endif
#  if defined(__SYMBIAN32__)
#    define XS_EXTERNAL(name) EXPORT_C XSPROTO(name)
#    define XS_INTERNAL(name) EXPORT_C STATIC XSPROTO(name)
#  endif
#  ifndef XS_EXTERNAL
#    if defined(HASATTRIBUTE_UNUSED) && !defined(__cplusplus)
#      define XS_EXTERNAL(name) void name(pTHX_ CV* cv __attribute__unused__)
#      define XS_INTERNAL(name) STATIC void name(pTHX_ CV* cv __attribute__unused__)
#    else
#      ifdef __cplusplus
#        define XS_EXTERNAL(name) extern "C" XSPROTO(name)
#        define XS_INTERNAL(name) static XSPROTO(name)
#      else
#        define XS_EXTERNAL(name) XSPROTO(name)
#        define XS_INTERNAL(name) STATIC XSPROTO(name)
#      endif
#    endif
#  endif
#endif

/* perl >= 5.10.0 && perl <= 5.15.1 */


/* The XS_EXTERNAL macro is used for functions that must not be static
 * like the boot XSUB of a module. If perl didn't have an XS_EXTERNAL
 * macro defined, the best we can do is assume XS is the same.
 * Dito for XS_INTERNAL.
 */
#ifndef XS_EXTERNAL
#  define XS_EXTERNAL(name) XS(name)
#endif
#ifndef XS_INTERNAL
#  define XS_INTERNAL(name) XS(name)
#endif

/* Now, finally, after all this mess, we want an ExtUtils::ParseXS
 * internal macro that we're free to redefine for varying linkage due
 * to the EXPORT_XSUB_SYMBOLS XS keyword. This is internal, use
 * XS_EXTERNAL(name) or XS_INTERNAL(name) in your code if you need to!
 */

#undef XS_EUPXS
#if defined(PERL_EUPXS_ALWAYS_EXPORT)
#  define XS_EUPXS(name) XS_EXTERNAL(name)
#else
   /* default to internal */
#  define XS_EUPXS(name) XS_INTERNAL(name)
#endif

#ifndef PERL_ARGS_ASSERT_CROAK_XS_USAGE
#define PERL_ARGS_ASSERT_CROAK_XS_USAGE assert(cv); assert(params)

/* prototype to pass -Wmissing-prototypes */
STATIC void
S_croak_xs_usage(pTHX_ const CV *const cv, const char *const params);

STATIC void
S_croak_xs_usage(pTHX_ const CV *const cv, const char *const params)
{
    const GV *const gv = CvGV(cv);

    PERL_ARGS_ASSERT_CROAK_XS_USAGE;

    if (gv) {
        const char *const gvname = GvNAME(gv);
        const HV *const stash = GvSTASH(gv);
        const char *const hvname = stash ? HvNAME(stash) : NULL;

        if (hvname)
            Perl_croak(aTHX_ "Usage: %s::%s(%s)", hvname, gvname, params);
        else
            Perl_croak(aTHX_ "Usage: %s(%s)", gvname, params);
    } else {
        /* Pants. I don't think that it should be possible to get here. */
        Perl_croak(aTHX_ "Usage: CODE(0x%"UVxf")(%s)", PTR2UV(cv), params);
    }
}
#undef  PERL_ARGS_ASSERT_CROAK_XS_USAGE

#ifdef PERL_IMPLICIT_CONTEXT
#define croak_xs_usage(a,b)    S_croak_xs_usage(aTHX_ a,b)
#else
#define croak_xs_usage        S_croak_xs_usage
#endif

#endif

/* NOTE: the prototype of newXSproto() is different in versions of perls,
 * so we define a portable version of newXSproto()
 */
#ifdef newXS_flags
#define newXSproto_portable(name, c_impl, file, proto) newXS_flags(name, c_impl, file, proto, 0)
#else
#define newXSproto_portable(name, c_impl, file, proto) (PL_Sv=(SV*)newXS(name, c_impl, file), sv_setpv(PL_Sv, proto), (CV*)PL_Sv)
#endif /* !defined(newXS_flags) */

#line 170 "SHEEP.c"

XS_EUPXS(XS_SHEEP_new); /* prototype to pass -Wmissing-prototypes */
XS_EUPXS(XS_SHEEP_new)
{
    dVAR; dXSARGS;
    if (items != 3)
       croak_xs_usage(cv,  "CLASS, _root, _shards");
    {
	char *	CLASS = (char *)SvPV_nolen(ST(0))
;
	Sheep *	RETVAL;
	char *	_root = (char *)SvPV_nolen(ST(1))
;
	int	_shards = (int)SvIV(ST(2))
;

	RETVAL = new Sheep(_root, _shards);
	ST(0) = sv_newmortal();
	sv_setref_pv( ST(0), CLASS, (void*)RETVAL );
    }
    XSRETURN(1);
}


XS_EUPXS(XS_SHEEP_DESTROY); /* prototype to pass -Wmissing-prototypes */
XS_EUPXS(XS_SHEEP_DESTROY)
{
    dVAR; dXSARGS;
    if (items != 1)
       croak_xs_usage(cv,  "THIS");
    {
	Sheep *	THIS;

	if( sv_isobject(ST(0)) && (SvTYPE(SvRV(ST(0))) == SVt_PVMG) )
		THIS = (Sheep *)SvIV((SV*)SvRV( ST(0) ));
	else{
		warn( "SHEEP::DESTROY() -- THIS is not a blessed SV reference" );
		XSRETURN_UNDEF;
	}
;

	delete THIS;
    }
    XSRETURN_EMPTY;
}


XS_EUPXS(XS_SHEEP_index); /* prototype to pass -Wmissing-prototypes */
XS_EUPXS(XS_SHEEP_index)
{
    dVAR; dXSARGS;
    if (items != 2)
       croak_xs_usage(cv,  "THIS, docs");
    {
	Sheep *	THIS;
	int	RETVAL;
	dXSTARG;
	AV *	docs;

	if( sv_isobject(ST(0)) && (SvTYPE(SvRV(ST(0))) == SVt_PVMG) )
		THIS = (Sheep *)SvIV((SV*)SvRV( ST(0) ));
	else{
		warn( "SHEEP::index() -- THIS is not a blessed SV reference" );
		XSRETURN_UNDEF;
	}
;

	STMT_START {
		SV* const xsub_tmp_sv = ST(1);
		SvGETMAGIC(xsub_tmp_sv);
		if (SvROK(xsub_tmp_sv) && SvTYPE(SvRV(xsub_tmp_sv)) == SVt_PVAV){
		    docs = (AV*)SvRV(xsub_tmp_sv);
		}
		else{
		    Perl_croak(aTHX_ "%s: %s is not an ARRAY reference",
				"SHEEP::index",
				"docs");
		}
	} STMT_END
;

	RETVAL = THIS->index(docs);
	XSprePUSH; PUSHi((IV)RETVAL);
    }
    XSRETURN(1);
}


XS_EUPXS(XS_SHEEP_search); /* prototype to pass -Wmissing-prototypes */
XS_EUPXS(XS_SHEEP_search)
{
    dVAR; dXSARGS;
    if (items != 3)
       croak_xs_usage(cv,  "THIS, query, n");
    {
	Sheep *	THIS;
	SV *	RETVAL;
	SV *	query = ST(1)
;
	int	n = (int)SvIV(ST(2))
;

	if( sv_isobject(ST(0)) && (SvTYPE(SvRV(ST(0))) == SVt_PVMG) )
		THIS = (Sheep *)SvIV((SV*)SvRV( ST(0) ));
	else{
		warn( "SHEEP::search() -- THIS is not a blessed SV reference" );
		XSRETURN_UNDEF;
	}
;

	RETVAL = THIS->search(query, n);
	ST(0) = RETVAL;
	sv_2mortal(ST(0));
    }
    XSRETURN(1);
}

#ifdef __cplusplus
extern "C"
#endif
XS_EXTERNAL(boot_SHEEP); /* prototype to pass -Wmissing-prototypes */
XS_EXTERNAL(boot_SHEEP)
{
    dVAR; dXSARGS;
#if (PERL_REVISION == 5 && PERL_VERSION < 9)
    char* file = __FILE__;
#else
    const char* file = __FILE__;
#endif

    PERL_UNUSED_VAR(cv); /* -W */
    PERL_UNUSED_VAR(items); /* -W */
#ifdef XS_APIVERSION_BOOTCHECK
    XS_APIVERSION_BOOTCHECK;
#endif
    XS_VERSION_BOOTCHECK;

        newXS("SHEEP::new", XS_SHEEP_new, file);
        newXS("SHEEP::DESTROY", XS_SHEEP_DESTROY, file);
        newXS("SHEEP::index", XS_SHEEP_index, file);
        newXS("SHEEP::search", XS_SHEEP_search, file);
#if (PERL_REVISION == 5 && PERL_VERSION >= 9)
  if (PL_unitcheckav)
       call_list(PL_scopestack_ix, PL_unitcheckav);
#endif
    XSRETURN_YES;
}



Makefile.PL
use 5.018001;
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
my $CC = 'g++';
WriteMakefile(
    NAME              => 'SHEEP',
    VERSION_FROM      => 'lib/SHEEP.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/SHEEP.pm', # retrieve abstract from module
       AUTHOR         => 'A. U. Thor <jack@nonet>') : ()),
    CC                => $CC,
    LD                => '$(CC)',
    CCFLAGS            => '-Wno-deprecated-declarations',
    LIBS              => ['-lclucene-core -lboost_thread -lboost_filesystem'], # e.g., '-lm'
    DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
    INC               => '-I.', # e.g., '-I. -I/usr/include/other'
	# Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
    'XSOPT'             => '-C++',
    'TYPEMAPS'          => ['perlobject.map' ],
);


cpp/Util.hpp
// from http://search.cpan.org/~tbusch/Lucene-0.01/lib/Lucene.pm

wchar_t* CharToWChar(int is_utf8, U8 *src, STRLEN arg_len) {
    wchar_t* ret;
    // Alloc memory for wide char string.  This could be a bit more
    // then necessary.
    Newz(0, ret, arg_len + 1, wchar_t);
    wchar_t* dst = ret;

    if (is_utf8) {
        // UTF8 to wide char mapping
        STRLEN len;
        while (*src) {
            *dst++ = utf8_to_uvuni(src, &len);
            src += len;
        }
    } else {
        // char to wide char mapping
        while (*src) {
            *dst++ = (wchar_t) *src++;
        }
    }
    *dst = 0;
    return ret;
}

wchar_t *SvToWChar(SV* arg) {
    wchar_t* ret;
    STRLEN arg_len;
    SvPV(arg, arg_len);
    return CharToWChar(SvUTF8(arg) ? 1 : 0,(U8 *) SvPV_nolen(arg),arg_len);
}

SV* WCharToSv(const wchar_t* src, SV* dest) {
    U8* dst;
    U8* d;

    // Alloc memory for wide char string.  This is clearly wider
    // then necessary in most cases but no choice.
    Newz(0, dst, 3 * wcslen(src) + 1, U8);

    d = dst;
    while (*src) {
        d = uvuni_to_utf8(d, *src++);
    }
    *d = 0;

    sv_setpv(dest, (char*) dst);
    sv_utf8_decode(dest);

    Safefree(dst);
    return dest;
}


cpp/Query.hpp
#include "CLucene.h"

using namespace lucene::index;
using namespace lucene::search;
using namespace lucene::analysis;
using namespace lucene::util;
using namespace lucene::store;
using namespace lucene::document;

BooleanQuery *add_query(BooleanQuery *parent, SV *query, BooleanClause::Occur occur);
BooleanQuery *add_queries_from_ar(BooleanQuery *parent, SV *_queries, BooleanClause::Occur occur);

BooleanQuery *add_queries_from_ar(BooleanQuery *parent, SV *_queries, BooleanClause::Occur occur) {
    AV *queries = (AV *)SvRV(_queries);
    int i;

    for (i=0; i<=av_len(queries); i++) {
        SV** elem = av_fetch(queries, i, 0);
        if (elem == NULL || *elem == &PL_sv_undef || !SvROK(*elem) || (SvTYPE(SvRV(*elem)) != SVt_PVHV))
            die("bool query expects hashref");

        add_query(parent,*elem,occur);
    }
    return parent;
}

BooleanQuery *add_query(BooleanQuery *parent, SV *query, BooleanClause::Occur occur) {
    HE *he;
    if (query == NULL || !SvROK(query) || (SvTYPE(SvRV(query)) != SVt_PVHV) )
        die("add_query expects hashref");

    HV *h_query = (HV *)SvRV(query);
    while (he = hv_iternext(h_query)) {
        I32 len;
        char *key = hv_iterkey(he,&len);
        SV *val = hv_iterval(h_query,he);
        if (val == &PL_sv_undef || val == NULL)
            die("undef in the query");
        if (!SvROK(val) || (SvTYPE(SvRV(val)) != SVt_PVHV) )
            die("expecting hashref");
        HV *h_val = (HV *)SvRV(val);
        HE *he_val;
        if (strncmp("term",key,len) == 0) {
            while (he_val = hv_iternext(h_val)) {
                I32 len;
                char *t_key = hv_iterkey(he_val,&len);
                SV *t_val = hv_iterval(h_val,he_val);
                wchar_t *w_key = CharToWChar(0,(U8 *) t_key,len);
                wchar_t *w_val = SvToWChar(t_val);
                Term *t = _CLNEW Term(w_key, w_val);
                parent->add(_CLNEW TermQuery(t),true,occur);
                Safefree(w_key);
                Safefree(w_val);
                _CLDECDELETE(t);
            }
        } else if (strncmp("bool",key,len) == 0) {
            SV **pp_should = hv_fetch(h_val,"should",6,0);
            SV **pp_must = hv_fetch(h_val,"must",4,0);
            SV **pp_must_not = hv_fetch(h_val,"must_not",8,0);

            #define PP_IS_NOT_AREF(elem) ((elem) != NULL && (!SvROK(*(elem)) || (SvTYPE(SvRV(*(elem))) != SVt_PVAV)) )
            if (PP_IS_NOT_AREF(pp_should) || PP_IS_NOT_AREF(pp_must) || PP_IS_NOT_AREF(pp_must_not))
                die("bool query expects { must => [], should => [], must_not => [] }");

            BooleanQuery *b = _CLNEW BooleanQuery();
            if (pp_should)
                add_queries_from_ar(b,*pp_should,BooleanClause::SHOULD);
            if (pp_must)
                add_queries_from_ar(b,*pp_must,BooleanClause::MUST);
            if (pp_must_not)
                add_queries_from_ar(b,*pp_must_not,BooleanClause::MUST_NOT);
            parent->add(b,true,occur);
        } else {
            die("dont know what to do with: %s",key);
        }
    }
    return parent;
}


cpp/Sheep.hpp
#include <iostream>
#include <boost/thread.hpp>
#include <boost/lexical_cast.hpp>
#include <boost/filesystem.hpp>
#include "CLucene.h"
#include "cpp/Util.hpp"
#include "cpp/Query.hpp"
namespace fs = boost::filesystem;

using namespace lucene::index;
using namespace lucene::search;
using namespace lucene::analysis;
using namespace lucene::util;
using namespace lucene::store;
using namespace lucene::document;
using namespace std;
#define MAX_SHARDS 128

class Sheep {
public:
    string root;
    int shards;
    int counter;
    FSDirectory *directories[MAX_SHARDS];
    lucene::analysis::WhitespaceAnalyzer an;

    Sheep(char * _root, int _shards) {
        if (_shards >= MAX_SHARDS)
            die("%d > MAX_SHARDS(%d)",_shards,MAX_SHARDS);

        this->root = string(_root);
        this->shards = _shards;
        this->counter = 0;
        int shard;
        for (shard = 0; shard < this->shards; shard++) {
            fs::path dir(this->root);
            dir /= (boost::lexical_cast<string>(shard));
            directories[shard] = FSDirectory::getDirectory(dir.c_str(), false, NULL);
        }
    }

    ~Sheep() {
        int shard;
        for (shard = 0; shard < this->shards; shard++) {
            if (this->directories[shard]) {
                this->directories[shard]->close();
                _CLLDELETE(this->directories[shard]);
            }
        }
    }

    void _search_into(Query *q, IndexSearcher *s, Hits **h) {
        try {
            *h = s->search(q);
        } catch (CLuceneError& e) {
            std::cout << e.what();
            *h = NULL;
        }
    }

    SV *search(SV *query, int n) {
        BooleanQuery *q = _CLNEW BooleanQuery();
        add_query(q,query,BooleanClause::MUST);
        IndexSearcher *searchers[this->shards];
        boost::thread t[this->shards];
        Hits *hits[this->shards];
        Document::FieldsType::const_iterator fit;
        const Document::FieldsType* fields;
        int shard;

        AV *ret = newAV();
        for (shard = 0; shard < this->shards; shard++) {
            searchers[shard] = _CLNEW IndexSearcher(directories[shard]);
            t[shard] = boost::thread(&Sheep::_search_into, this, q,searchers[shard],&hits[shard]);
        }

        for (shard = 0; shard < this->shards; shard++) {
            t[shard].join();

            if (!hits[shard]) {
                av_undef(ret);
                die("search error: look at %s:%d, there should also be an error in stdout",__FILE__,__LINE__);
            }
            for (size_t i = 0; i < std::min((size_t )n,hits[shard]->length()); i++ ){
                Document* doc = &hits[shard]->doc(i);
                fields = doc->getFields();
                HV *item = newHV();
                fit = fields->begin();
                while (fit != fields->end()) {
                    Field* field = *fit;
                    SV *f_val = newSV(0);
                    SV *f_name = newSV(0);
                    WCharToSv(field->name(),f_name);
                    WCharToSv((const TCHAR *)field->stringValue(),f_val);
                    hv_store_ent(item,f_name,f_val,0);

                    fit++;
                }
                hv_store(item,"_score",6, newSVnv(hits[shard]->score(i)),0);
                av_push(ret,newRV_noinc((SV *)item));
            }
            _CLLDELETE(hits[shard]);
            searchers[shard]->close();
            _CLLDELETE(searchers[shard]);
        }

        _CLLDELETE(q);
        return newRV_noinc((SV *)ret);
    }

    int index(AV *docs) {
        int i,inserted = 0, shard;
        try {
            IndexWriter *writers[this->shards];
            for (shard = 0; shard < this->shards; shard++) {
                bool should_create = IndexReader::indexExists(this->directories[shard]) ? false : true;
                writers[shard] = _CLNEW IndexWriter(this->directories[shard],&an, should_create);
                writers[shard]->setUseCompoundFile(false);
            }

            Document doc;
            HE *he;
            for (i=0; i<= av_len(docs); i++) {
                SV** elem = av_fetch(docs, i, 0);
                if (elem == NULL || *elem == &PL_sv_undef || !SvROK(*elem) || (SvTYPE(SvRV(*elem)) != SVt_PVHV) )
                    continue;

                HV *hv = (HV *)SvRV(*elem);
                doc.clear();
                while (he = hv_iternext(hv)) {
                    I32 len;
                    char *key = hv_iterkey(he,&len);
                    SV *val = hv_iterval(hv,he);
                    wchar_t *w_key = CharToWChar(0,(U8 *) key,len);
                    wchar_t *w_val = SvToWChar(val);
                    doc.add(*_CLNEW Field(w_key, w_val, Field::STORE_YES | Field::INDEX_TOKENIZED) );
                    Safefree(w_key);
                    Safefree(w_val);
                }
                writers[this->counter++ % this->shards]->addDocument( &doc );
                inserted++;
            }
            for (shard = 0; shard < this->shards; shard++) {
                writers[shard]->setUseCompoundFile(true);
                writers[shard]->optimize();
                writers[shard]->close();
                _CLLDELETE(writers[shard]);
            }
        } catch (CLuceneError& e) {
            die(e.what());
        }
        return inserted;
    }
};


lib/SHEEP.pm
package SHEEP;

use 5.018001;
use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# Items to export into callers namespace by default. Note: do not export
# names by default without a very good reason. Use EXPORT_OK instead.
# Do not simply export all your public functions/methods/constants.

# This allows declaration	use SHEEP ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = ( 'all' => [ qw(
	
) ] );

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(

);

our $VERSION = '0.01';

require XSLoader;
XSLoader::load('SHEEP', $VERSION);

# Preloaded methods go here.

1;
__END__
# Below is stub documentation for your module. You'd better edit it!
=head1 ALPHA - WORK IN PROGRESS

*alpha* not recommended for anything but hacking on some simple CLucene examples

=head1 NAME

SHEEP - simple CLucene wrapper that spawns threads for every shard when search()ing

=head1 SYNOPSIS

  use SHEEP;
  my $x = new SHEEP('/path/to/index',5) # create index with 5 shards

  my @documents = ();
  push @documents, { "name" => "jack doe", age => "200" }
      for(1 .. 1_000);

  push @documents, { "name" => "john doe", age => "300" }
      for(1 .. 1_000);
  $x->index(\@documents);

  # find top 10 documents containing 'doe' in the name field
  my $r = $x->search({ term => { name => 'doe' } }, 10); 

  # find top 10 documents containing 'doe' in the name,
  # AND 200 in the 'age' field
  my $r = $x->search({ term => { name => 'doe', age => '200' } }, 10); 

  # some bool query examples:
  $x->search({ 
                bool => {
                    must => [
                       { term => { "name" => 'jack' } }
                    ],
                    must_not => [
                       { term => { "name" => 'john' } }
                    ]
                }
             }, 10);

  $x->search({ 
                bool => {
                    must => [
                       { term => { "name" => 'jack' } }
                    ],
                    must_not => [
                       { term => { "name" => 'jack' } }
                    ]
                }
             }, 10);

  $x->search({ 
                bool => {
                    must => [
                       { term => { "name" => 'jack' } }
                    ],
                    should => [
                       { term => { "name" => 'jim' } }
                    ],
                    must_not => [
                       { term => { "name" => 'john' } }
                    ]
                }
             }, 10);

  $x->search({ 
                bool => {
                    must => [
                       { 
                           bool => {  
                               must => [ { term => { name => 'jack' } } ]
                           }
                       }
                    ],
                    should => [
                       { term => { name => 'jim' } }
                    ],
                    must_not => [
                       { term => { name => 'john' } }
                    ]
                }
             }, 10);



=head1 DESCRIPTION

SHEEP is a simplistic non-flexible CLucene wrapper

It uses WhiteSpaceAnalyzer, and can work only with BooleanQueries and Term queries. It also uses boost::thread to spawn new thread for every shard when searching.

=head1 SEE ALSO

L<http://clucene.sourceforge.net/>

L<http://clucene.sourceforge.net/doc/html/namespaces.html>

L<https://github.com/jackdoe/SHEEP>

=head1 AUTHOR

Borislav Nikolov <jack@sofialondonmoskva.com>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2013 by Borislav Nikolov

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.18.1 or,
at your option, any later version of Perl 5 you may have available.


=cut
